File: scripts/check-product-config.py
from __future__ import annotations

from collections import Counter
from pathlib import Path
from typing import Any, Dict, List

try:
    import yaml
except ModuleNotFoundError:
    yaml = None


ROOT = Path(__file__).resolve().parents[1]
PRODUCT_CONFIG = ROOT / "config" / "driver-products.yml"

BASE_REQUIRED = {
    "status",
    "source",
    "gee_id",
    "type",
    "source_temporal_resolution",
    "output_temporal_resolution",
    "output_name",
}

CONTINUOUS_REQUIRED = {
    "band",
    "output_units",
    "scale_factor",
    "offset",
    "reducer",
}


def load_products() -> Dict[str, Any]:
    if yaml is not None:
        with PRODUCT_CONFIG.open(encoding="utf-8") as handle:
            config = yaml.safe_load(handle) or {}

        return config.get("products") or {}

    return load_products_without_yaml()


def parse_scalar(value: str) -> Any:
    if value in {"true", "True"}:
        return True
    if value in {"false", "False"}:
        return False
    if value in {"null", "None", ""}:
        return None

    try:
        if "." in value:
            return float(value)
        return int(value)
    except ValueError:
        return value


def load_products_without_yaml() -> Dict[str, Any]:
    products: Dict[str, Any] = {}
    current_product = None
    in_products = False

    for raw_line in PRODUCT_CONFIG.read_text(encoding="utf-8").splitlines():
        if not raw_line.strip() or raw_line.lstrip().startswith("#"):
            continue

        indent = len(raw_line) - len(raw_line.lstrip(" "))
        line = raw_line.strip()

        if indent == 0 and line == "products:":
            in_products = True
            continue

        if not in_products:
            continue

        if indent == 2 and line.endswith(":"):
            current_product = line[:-1]
            products[current_product] = {}
            continue

        if indent == 4 and current_product and ":" in line:
            key, value = line.split(":", 1)
            products[current_product][key.strip()] = parse_scalar(value.strip())

    return products


def missing_fields(product: Dict[str, Any]) -> List[str]:
    required = set(BASE_REQUIRED)

    if product.get("type") == "continuous":
        required.update(CONTINUOUS_REQUIRED)

    if product.get("type") != "source_list" and product.get("source_temporal_resolution") != "static":
        required.add("selected_spatial_resolution_m")

    return sorted(field for field in required if field not in product)


def main() -> None:
    products = load_products()
    issues = []

    for name, settings in products.items():
        missing = missing_fields(settings or {})
        if missing:
            issues.append((name, missing))

    status_counts = Counter((settings or {}).get("status", "missing") for settings in products.values())
    type_counts = Counter((settings or {}).get("type", "missing") for settings in products.values())

    print(f"Products: {len(products)}")
    print("By status:", dict(sorted(status_counts.items())))
    print("By type:", dict(sorted(type_counts.items())))

    if issues:
        print("\nMissing fields:")
        for name, missing in issues:
            print(f"- {name}: {', '.join(missing)}")
        raise SystemExit(1)

    print("\nProduct config looks OK")


if __name__ == "__main__":
    main()

File: scripts/check-product-config.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(yaml)
})

`%||%` <- function(x, y) {
  if (is.null(x)) y else x
}

script_arg <- grep("^--file=", commandArgs(FALSE), value = TRUE)
script_path <- if (length(script_arg)) sub("^--file=", "", script_arg[[1]]) else "scripts/check-product-config.R"
repo_root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)
product_config <- file.path(repo_root, "config", "driver-products.yml")

base_required <- c(
  "status",
  "source",
  "gee_id",
  "type",
  "source_temporal_resolution",
  "output_temporal_resolution",
  "output_name"
)

continuous_required <- c(
  "band",
  "output_units",
  "scale_factor",
  "offset",
  "reducer"
)

missing_fields <- function(product) {
  required <- base_required

  if (identical(product$type, "continuous")) {
    required <- c(required, continuous_required)
  }

  if (!identical(product$type, "source_list") &&
      !identical(product$source_temporal_resolution, "static")) {
    required <- c(required, "selected_spatial_resolution_m")
  }

  sort(setdiff(unique(required), names(product)))
}

config <- yaml::read_yaml(product_config)
products <- config$products

if (is.null(products) || !length(products)) {
  stop("No products found in config/driver-products.yml.", call. = FALSE)
}

issues <- lapply(names(products), function(name) {
  missing <- missing_fields(products[[name]] %||% list())
  if (!length(missing)) {
    return(NULL)
  }
  list(name = name, missing = missing)
})
issues <- Filter(Negate(is.null), issues)

status_values <- vapply(products, function(product) product$status %||% "missing", character(1))
type_values <- vapply(products, function(product) product$type %||% "missing", character(1))

cat("Products:", length(products), "\n")
cat("By status:", paste(names(sort(table(status_values))), as.integer(sort(table(status_values))), sep = "=", collapse = ", "), "\n")
cat("By type:", paste(names(sort(table(type_values))), as.integer(sort(table(type_values))), sep = "=", collapse = ", "), "\n")

if (length(issues)) {
  cat("\nMissing fields:\n")
  for (issue in issues) {
    cat("- ", issue$name, ": ", paste(issue$missing, collapse = ", "), "\n", sep = "")
  }
  quit(status = 1)
}

cat("\nProduct config looks OK\n")
